source/common/router/router.h

```diff
@@ -138,6 +145,32 @@ class Filter : Logger::Loggable<Logger::Id::router>,
     return callbacks_->connection();
   }
 
+  /**
+   * Set a pseudo-random cookie with the name @param key and TTL @param max_age
+   * to be sent to the downstream host.
+   * @return std::string the value of the new cookie
+   *
+   * marked const so it can be called from hashKey(), and because all mutated
+   * state is only visible when calling onUpstreamHeaders()
+   */
+  std::string addDownstreamSetCookie(const std::string& key, std::chrono::seconds max_age) const {
+    // The cookie value should be the same per connection so that if multiple
+    // streams race on the same path, they all receive the same cookie.
```

Comment:
As discussed, in person I'm not quite sure what the right thing to do here is on path diversity vs hashing consistency.

If most Envoy servers were set up with HTTP browsers hitting a single Envoy it'd be nice to hash on IP rather than Address to have consistent hashing for the 8 HTTP connections.  Unfortunately that greatly reduces hash diversity so if you instead have a lot of connections from one source to one dest they'd end up all pointed at the same place.

I'd be inclined to err on the side of not overloading backends, but maybe worth a comment that for the HTTP + multiple connections case there will still be a cookie race.

Reply:
Are we going to want to go to a harder form of sticky further down the track? If so, then some of this will be moot.

Reply:
I think this is part of the convo above, which I think partly happened IRL, but, I'm a little confused as to why the local address is part of the hash. Also related to stickiness per @htuch. Can you maybe add some more detailed comments here on the rationale behind how this is being set / the overall algorithm being used?